sklearn/decomposition/_nmf.py

```diff
@@ -1304,24 +1246,97 @@ def fit_transform(self, X, y=None, W=None, H=None):
         W : ndarray of shape (n_samples, n_components)
             Transformed data.
         """
-        X = self._validate_data(X, accept_sparse=('csr', 'csc'),
-                                dtype=[np.float64, np.float32])
+        n_components = self.n_components
+        init = self.init
+        solver = self.solver
+        beta_loss = self.beta_loss
+        tol = self.tol
+        max_iter = self.max_iter
+        random_state = self.random_state
+        alpha = self.alpha
+        l1_ratio = self.l1_ratio
+        verbose = self.verbose
+        shuffle = self.shuffle
+        regularization = self.regularization
+        update_H = self.update_H
+
+        X = check_array(X, accept_sparse=('csr', 'csc'),
+                        dtype=[np.float64, np.float32])
+        check_non_negative(X, "NMF (input X)")
+        beta_loss = _check_string_param(solver, regularization,
+                                        beta_loss, init)
+
+        if X.min() == 0 and beta_loss <= 0:
+            raise ValueError("When beta_loss <= 0 and X contains zeros, "
+                             "the solver may diverge. Please add small values "
+                             "to X, or use a positive beta_loss.")
+
+        n_samples, n_features = X.shape
+        if n_components is None:
+            n_components = n_features
+
+        if not isinstance(n_components, numbers.Integral) or n_components <= 0:
+            raise ValueError("Number of components must be a positive integer;"
+                             " got (n_components=%r)" % n_components)
+        if not isinstance(max_iter, numbers.Integral) or max_iter < 0:
+            raise ValueError("Maximum number of iterations must be a positive "
+                             "integer; got (max_iter=%r)" % max_iter)
+        if not isinstance(tol, numbers.Number) or tol < 0:
+            raise ValueError("Tolerance for stopping criteria must be "
+                             "positive; got (tol=%r)" % tol)
```

Comment:
I would move all these param validation in a dedicated private method, ``_check_params``, to make the code more readable and more reusable by MiniBatchNMF